File: Exercises2024/Chapter09_various_ensembles/Fortran/3-IsingModel/Source/ising.f
      Program Ising
      Implicit None

Ccccccccccccccccccccccccccc
C     2d Ising Model      C
C     No Pbc              C
C     4 Neighbors         C
Ccccccccccccccccccccccccccc

      Integer Maxlat,LatticeSize,Iup(4),Jup(4),Magnet,Energy,
     &     I,J,Inew,Jnew,Lnew,Lold,Diff,Sstmm,Ilat,Jlat,
     &     K,L,NumberOfCycles,Ninit,Maxmag,Mnew,Mold
      Integer CycleMultiplication

      Parameter (Maxlat = 32   )
      Parameter (Maxmag = 1024 )
      Parameter (CycleMultiplication = 100 )

      Integer Lattice(0:Maxlat+1,0:Maxlat+1)

      Double Precision M1,RandomNumber,Beta,EnergySum,
     &     Dist2,Weight,Accepted,Attempts,Dist3,EnergyCount,
     &     W(0:Maxmag),Distribution(-Maxmag:Maxmag),
     &     Ttime,Tstart

Ccccccccccccccccccccccccc
C     Initialize Rng    C
Ccccccccccccccccccccccccc

      M1 = 0.001d0*Dble(Mod((10+10*Sstmm()),1000))

      If(M1.Lt.0.001d0) M1 = 0.001d0
      If(M1.Gt.0.999d0) M1 = 0.999d0

      Call InitializeRandomNumberGenerator(M1) 
 
Ccccccccccccccccccccccccc
C     Read Info         C
Ccccccccccccccccccccccccc

      Write(6,*) '2d Ising Model, 4 Neightbors, No Pbc'
      Write(6,*)
      Write(6,*) 'Written By Thijs J.H. Vlugt Nov. 1999'
      Write(6,*)
      Write(6,*)

      Write(6,*) 'Lattice Size                     ? '
      Read(*,*) LatticeSize

      If(Mod(LatticeSize,2).Ne.0.Or.Mod(Maxmag,2).Ne.0.Or.
     &     Mod(Maxlat,2).Ne.0) Stop
      
      Write(6,*) 'Beta                             ? '
      Read(*,*) Beta

      Write(6,*) 'Number Of Cycles                 ? '
      Read(*,*) NumberOfCycles

      If(LatticeSize.Lt.3.Or.LatticeSize.Gt.Maxlat) Stop
      If(NumberOfCycles.Lt.10) Stop

      Do I=0,Maxmag
         W(I) = 1.0d0
      Enddo
      
Cccccccccccccccccccccccccccccccccc
C     Multicanonical Algorithm   C
Cccccccccccccccccccccccccccccccccc

      Open(21,File="w.dat")

      Do I=0,Maxmag,2
         Read(21,*) K,EnergySum
         
         If(Abs(K).Gt.Maxmag.Or.Mod(K,2).Ne.0) Then
            Write(6,*) 'I Out Of Range !!'
            Stop
         Endif

         W(K) = EnergySum
      Enddo

      Close(21)
      
      Ninit = NumberOfCycles/3
      EnergySum = 0.0d0
      EnergyCount = 0.0d0
      Accepted = 0.0d0
      Attempts = 0.0d0
      Tstart = Ttime()
      
      If(Ninit.Gt.100) Ninit = 100

      Do I=-Maxmag,Maxmag
         Distribution(I) = 0.0d0
      Enddo

Cccccccccccccccccccccccccccccccccccccccccccccc
C     Initialize Lattice                     C
C     Random Lattice                         C
C                                            C
C     Add One Boundary Layer With Spin=0     C
C     This Is A Way To Avoid If-Statements   C
C     In The Computation Of The Energy       C
C                                            C
C     LatticeSize  = Size Of The Lattice     C
C     Lattice(I,J) = Spin Of Site (I,J)      C
Cccccccccccccccccccccccccccccccccccccccccccccc

      Do I=0,Maxlat+1
         Do J=0,Maxlat+1
            Lattice(I,J) = 0
         Enddo
      Enddo

      Do I=1,LatticeSize
         Do J=1,LatticeSize
            If(RandomNumber().Lt.0.5d0) Then
               Lattice(I,J) =  1
            Else
               Lattice(I,J) = -1
            Endif
         Enddo
      Enddo

Ccccccccccccccccccccccccccccccc
C     Initialize Neighbours   C
Ccccccccccccccccccccccccccccccc

      Iup(1) =  1
      Jup(1) =  0
      
      Iup(2) = -1
      Jup(2) =  0

      Iup(3) =  0
      Jup(3) =  1
      
      Iup(4) =  0
      Jup(4) = -1

Ccccccccccccccccccccccccccccccccccccccc
C     Calculate Initial Energy        C
C                                     C
C     Magnet = Total Magnetisation    C
Ccccccccccccccccccccccccccccccccccccccc

      Energy = 0
      Magnet = 0

      Do I=1,LatticeSize
         Do J=1,LatticeSize

            Magnet = Magnet + Lattice(I,J)

            Do K=1,4
               Inew   = I + Iup(K)
               Jnew   = J + Jup(K)
               Energy = Energy - 
     &              Lattice(I,J)*Lattice(Inew,Jnew)
            Enddo
         Enddo
      Enddo

      Energy = Energy/2

      Write(6,*)
      Write(6,*) 'Lattice Size               : ',LatticeSize
      Write(6,*) 'Beta                       : ',Beta
      Write(6,*) 'Number Of Cycles           : ',NumberOfCycles
      Write(6,*) 'Number Of Init             : ',Ninit
      Write(6,*)
      Write(6,*)
      Write(6,*) 'Initial Energy             : ',Energy
      Write(6,*) 'Initial Magnetization      : ',Magnet
      Write(6,*)
      
      If(Abs(Magnet).Gt.Maxmag) Stop

Cccccccccccccccccccccccccccccc
C     Loop Over All Cycles   C
Cccccccccccccccccccccccccccccc

      Do K=1,NumberOfCycles
         Do L=1,CycleMultiplication*LatticeSize*LatticeSize
            
Ccccccccccccccccccccccccccccccccc
C     Flip A Single Spin        C
C     Metropolis Algorithm      C
Ccccccccccccccccccccccccccccccccc

            Ilat = 1 + Int(RandomNumber()*Dble(LatticeSize))
            Jlat = 1 + Int(RandomNumber()*Dble(LatticeSize))
            
            Mold  = Magnet
            Lold  = Lattice(Ilat,Jlat)
            Lnew  = -Lold
            Mnew  = Mold + Lnew - Lold
            Diff  = 0
            Attempts = Attempts + 1.0d0
            
Cccccccccccccccccccccccccccccccccccccccccc
C     Calculate The Energy Difference    C
C     Between New And Old                C
Cccccccccccccccccccccccccccccccccccccccccc

C     Start Modification

C     End Modification

CCcccccccccccccccccccccccccccccccccc
C     Acceptance/Rejection Rule    C
C     Use Weight Function W        C
Cccccccccccccccccccccccccccccccccccc

            If(RandomNumber().Lt.
     &           Dexp((-Beta*Dble(Diff)) + 
     &           W(Abs(Mnew)) - W(Abs(Mold)))) Then

Cccccccccccccccccccccccccccccccccccccccccccccccccc
C     Update The Lattice/Energy/Magnetisation    C
Cccccccccccccccccccccccccccccccccccccccccccccccccc

               Accepted = Accepted + 1.0d0

C     Start Modification

C     End Modification
               
               If(Abs(Magnet).Gt.Maxmag) Then
                  Write(6,*) 'Increase The Value Of Maxmag !!'
                  Stop
               Endif
            Endif
                  
            If(K.Gt.Ninit) Then
               Weight = Dexp(-W(Abs(Magnet)))
               EnergySum = EnergySum + Weight*Dble(Energy)
               EnergyCount = EnergyCount + Weight
               
               Distribution(Magnet) = 
     &              Distribution(Magnet) + 1.0d0
            Endif
         Enddo
      Enddo

      Write(6,*)
      Write(6,*) 'Average Energy             : ',
     &     EnergySum/EnergyCount

      Write(6,*) 'Fraction Accepted Swaps    : ',
     &     Accepted/Attempts

      Write(6,*) 'Elapsed Time [S]           : ',
     &     Ttime()-Tstart

      Write(6,*)
      Write(6,*) 'Final Energy        (Simu) : ',Energy
      Write(6,*) 'Final Magnetization (Simu) : ',Magnet

Cccccccccccccccccccccccccccccccccc
C     Calculate Distributions    C
C     Magnetisation              C
Cccccccccccccccccccccccccccccccccc

      Dist2 = 0.0d0
      Dist3 = 0.0d0
            
      Do I=-Maxmag,Maxmag,2
         Weight = Dexp(-W(Abs(I)))

         Dist2  = Dist2 + Distribution(I)
         Dist3  = Dist3 + Distribution(I)*Weight
      Enddo

      Open(21,File="magnetic.dat")

      Do I=-Maxmag,Maxmag,2
         If(Distribution(I).Gt.0.5d0) Then
            Weight = Dexp(-W(Abs(I)))
            
            Write(21,*) I,Distribution(I)/Dist2,
     &           Distribution(I)/Dist3*Weight
         Endif
      Enddo   

      Close(21)

Ccccccccccccccccccccccccccccc
C     Weight Distribution   C
Ccccccccccccccccccccccccccccc

      Dist2 = 0.0d0

      Do I=0,Maxmag,2
         Distribution(I) = Max(1.0d0,Distribution(I) + Distribution(-I))*
     &        Dexp(-W(Abs(I)))

         Distribution(I) = -Dlog(Distribution(I))
         Dist2 = Min(Dist2,Distribution(I))
      Enddo

      Do I=0,Maxmag,2
         Distribution(I) = Distribution(I) - Dist2
         Distribution(I) = Min(10.0d0,1.05d0*Distribution(I))
      Enddo

      Open(21,File="w.dat")

      Do I=0,Maxmag,2
         Write(21,*) I,Distribution(I)
      Enddo

      Close(21)
      
Cccccccccccccccccccccccccccccccccc
C     Calculate Final Energy     C
C     Used To Check The Code     C
Cccccccccccccccccccccccccccccccccc

      Energy = 0
      Magnet = 0

      Do I=1,LatticeSize
         Do J=1,LatticeSize

            Magnet = Magnet + Lattice(I,J)

            Do K=1,4
               Inew   = I + Iup(K)
               Jnew   = J + Jup(K)
               Energy = Energy - 
     &              Lattice(I,J)*Lattice(Inew,Jnew)
            Enddo
         Enddo
      Enddo

      Energy = Energy/2

      Write(6,*) 'Final Energy        (Calc) : ',Energy
      Write(6,*) 'Final Magnetization (Calc) : ',Magnet
      Write(6,*)     

      Stop
      End

File: Exercises2024/Chapter09_various_ensembles/C/3-IsingModel/Source/ising.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>
#include "ran_uniform.h"

#define SQR(x) ((x)*(x))
#define CycleMultiplication 100

// 2d Ising Model
// No Periodic boundary conditions
// 4 Neighbors 

#define MAX_LATTICE_SIZE 32
#define Maxmag 1024
#define MAX_MAGNETIZATION 1024

#define MIN(x,y) ((x)<(y)?(x):(y))
#define MAX(x,y) ((x)>(y)?(x):(y))

int main(void)
{
  int i,j,k,l;
  int NumberOfCycles,NumberOfInitializationSteps;
  int Magnetization,LatticeSize;
  int Inew,Jnew,Lnew,Lold,Diff,Ilat,Jlat;
  int Mnew,Mold;
  double Energy,Beta,Av1,Dist2,Weight,Move1,Move2,Dist3,Av2;
  double W[MAX_MAGNETIZATION+1],Dist1[2*MAX_MAGNETIZATION+1];
  int Lattice[MAX_LATTICE_SIZE+2][MAX_LATTICE_SIZE+2];
  int Iup[4]={1,-1,0,0};
  int Jup[4]={0,0,1,-1};
  FILE *FilePtr;

  InitializeRandomNumberGenerator(time(0l)) ;
 
  // read info 
  printf("2d Ising Model, 4 Neightbors, No Pbc\n\n");

  printf("Lattice Size                     ? ");
  fscanf(stdin,"%d",&LatticeSize);

  if(LatticeSize%2!=0||MAX_MAGNETIZATION%2!=0||MAX_LATTICE_SIZE%2!=0)
  {
    printf("error in input parameters\n");
    exit(1);
  }
      
  printf("Beta                             ? ");
  fscanf(stdin,"%lf",&Beta);

  printf("Number Of Cycles                 ? ");
  fscanf(stdin,"%d",&NumberOfCycles);

  if(LatticeSize<3||LatticeSize>MAX_LATTICE_SIZE||NumberOfCycles<10)
  {
    printf("error in input parameters\n");
    exit(1);
  }

  for(i=0;i<MAX_MAGNETIZATION;i++)
    W[i]=1.0;
      
  // multicanonical algorithm 
  FilePtr=fopen("w.dat","r");
  for(i=0;i<MAX_MAGNETIZATION;i+=2)
  {
    fscanf(FilePtr,"%d %lf\n",&j,&Av1);
    if(abs(j)>MAX_MAGNETIZATION||j%2!=0)
    {
      printf("j out of range\n");
      exit(1);
    }
    W[j]=Av1;
  }
  fclose(FilePtr);
      
  NumberOfInitializationSteps=NumberOfCycles/3;
  Av1=0.0;
  Av2=0.0;
  Move1=0.0;
  Move2=0.0;
      
  if(NumberOfInitializationSteps>100) NumberOfInitializationSteps=100;

  for(i=0;i<2*MAX_MAGNETIZATION+1;i++)
    Dist1[i]=0.0;

  // Initialize Lattice
  // Random Lattice
  // 
  // add one boundary layer with spin=0
  // this is a way to avoid if-statements
  // in the computation of the energy

  for(i=0;i<=MAX_LATTICE_SIZE+1;i++)
    for(j=0;j<=MAX_LATTICE_SIZE+1;j++)
      Lattice[i][j]=0;

  for(i=1;i<=LatticeSize;i++)
    for(j=1;j<=LatticeSize;j++)
    {
      if(RandomNumber()<0.5)
        Lattice[i][j]=1;
      else
        Lattice[i][j]=-1;
    }

  // calculate initial energy
  Energy=0.0;
  Magnetization=0.0;

  for(i=1;i<=LatticeSize;i++)
    for(j=1;j<=LatticeSize;j++)
    {
      Magnetization+=Lattice[i][j];

      for(k=0;k<4;k++)
      {
        Inew=i+Iup[k];
        Jnew=j+Jup[k];
        Energy-=0.5*Lattice[i][j]*Lattice[Inew][Jnew];
      }
    }

    printf("\n");
    printf("Lattice Size               : %d\n",LatticeSize);
    printf("Beta                       : %f\n",Beta);
    printf("Number Of Cycles           : %d\n",NumberOfCycles);
    printf("Number Of Init             : %d\n",NumberOfInitializationSteps);
    printf("\n");
    printf("\n");
    printf("Initial Energy             : %f\n",Energy);
    printf("Initial Magnetization      : %d\n",Magnetization);
    printf("\n");
      
    if(fabs(Magnetization)>MAX_MAGNETIZATION)
    {
      printf("|Magnetization|>MAX_MAGNETIZATION\n");
      exit(0);
    }

  // loop over all cycles
  for(l=0;l<NumberOfCycles;l++)
  {
    for(k=0;k<CycleMultiplication*SQR(LatticeSize);k++)
    {
      // flip a single spin
      // Metropolis algorithm
      Ilat=1+(int)(RandomNumber()*LatticeSize);
      Jlat=1+(int)(RandomNumber()*LatticeSize);
            
      Mold=Magnetization;
      Lold=Lattice[Ilat][Jlat];
      Lnew=-Lold;
      Mnew=Mold+Lnew-Lold;
      Diff=0;
      Move2=Move2+1.0;
            
      // calculate the energy difference
      // between new and old

      // start modification

      // end modification

      // acceptance/rejection rule
      // use weight function W 
      if(RandomNumber()<exp(-Beta*Diff+W[abs(Mnew)]-W[abs(Mold)]))
      {
        // update the lattice/energy/magnetisation
        Move1=Move1+1.0;

        // start modification

        // end modification
               
        if(fabs(Magnetization)>MAX_MAGNETIZATION)
        {
          printf("Increase the value of maxmag\n");
          exit(0);
        }
      }
                  
      if(l>NumberOfInitializationSteps)
      {
        Weight=exp(-W[abs(Magnetization)]);
        Av1=Av1+Weight*Energy;
        Av2=Av2+Weight;
        Dist1[Magnetization+MAX_MAGNETIZATION]+=1.0;
      }
    }
  }

  printf("\n");
  printf("Average Energy             : %f\n",Av1/Av2);
  printf("Fraction Accepted Swaps    : %f\n",Move1/Move2);
  printf("\n");
  printf("Final Energy        (Simu) : %f\n",Energy);
  printf("Final Magnetization (Simu) : %d\n",Magnetization);
  printf("\n");

  // calculate distributions magnetization
  Dist2=0.0;
  Dist3=0.0;
            
  for(i=-MAX_MAGNETIZATION;i<=MAX_MAGNETIZATION;i+=2)
  {
    Weight=exp(-W[abs(i)]);
    Dist2+=Dist1[i+MAX_MAGNETIZATION];
    Dist3+=Dist1[i+MAX_MAGNETIZATION]*Weight;
  }

  FilePtr=fopen("magnetic.dat","w");
  for(i=-MAX_MAGNETIZATION;i<=MAX_MAGNETIZATION;i+=2)
  {
    if(Dist1[i+MAX_MAGNETIZATION]>0.5)
    {
      Weight=exp(-W[abs(i)]);
      fprintf(FilePtr,"%d %f %f\n",i,Dist1[i+MAX_MAGNETIZATION]/Dist2,Dist1[i+MAX_MAGNETIZATION]*Weight/Dist3);
    }
  }
  fclose(FilePtr);

  // weight distribution
  Dist2=0.0;
  for(i=0;i<=MAX_MAGNETIZATION;i+=2)
  {
    Dist1[i+MAX_MAGNETIZATION]=-log(MAX(1.0,Dist1[i+MAX_MAGNETIZATION]+Dist1[-i+MAX_MAGNETIZATION])*exp(-W[abs(i)]));
    Dist2=MIN(Dist2,Dist1[i+MAX_MAGNETIZATION]);
  }

  for(i=0;i<=MAX_MAGNETIZATION;i+=2)
  {
    Dist1[i+MAX_MAGNETIZATION]=Dist1[i+MAX_MAGNETIZATION]-Dist2;
    Dist1[i+MAX_MAGNETIZATION]=MIN(10.0,1.05*Dist1[i+MAX_MAGNETIZATION]);
  }

  FilePtr=fopen("w.dat","w");
  for(i=0;i<=MAX_MAGNETIZATION;i+=2)
    fprintf(FilePtr,"%d %f\n",i,Dist1[i+MAX_MAGNETIZATION]);
  fclose(FilePtr);
      
  // calculate final energy
  // used to check the code
  Energy=0;
  Magnetization=0;
  for(i=1;i<=LatticeSize;i++)
    for(j=1;j<=LatticeSize;j++)
    {
      Magnetization+=Lattice[i][j];
      for(k=0;k<4;k++)
      {
        Inew=i+Iup[k];
        Jnew=j+Jup[k];
        Energy-=0.5*Lattice[i][j]*Lattice[Inew][Jnew];
      }
    }

  printf("Final Energy        (Calc) : %f\n",Energy);
  printf("Final Magnetization (Calc) : %d\n",Magnetization);

  return 0;
}
